certs/trustpin.go

```diff
@@ -0,0 +1,90 @@
+package certs
+
+import (
+	"crypto/x509"
+	"fmt"
+	"github.com/docker/notary"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/utils"
+	"strings"
+)
+
+// TrustPinChecker handles logic for trust pinning for a gun, given a TrustPinConfig
+type TrustPinChecker struct {
+	mode         TrustPinMode
+	gun          string
+	config       notary.TrustPinConfig
+	pinnedCAPool *x509.CertPool
+}
+
+// TrustPinMode is a type to distinguish possible trust pinning modes
+type TrustPinMode string
+
+const (
+	tofus TrustPinMode = "tofu"
+	ca    TrustPinMode = "ca"
+	certs TrustPinMode = "certs"
+)
+
+// NewTrustPinChecker returns a new TrustPinChecker from a TrustPinConfig for a GUN
+func NewTrustPinChecker(trustPinConfig notary.TrustPinConfig, gun string) (TrustPinChecker, error) {
+	trustPinChecker := TrustPinChecker{config: trustPinConfig, gun: gun}
+	// Determine the mode, and if it's even valid
+	if _, ok := trustPinConfig.Certs[gun]; ok {
+		trustPinChecker.mode = certs
+	} else if utils.ContainsKeyPrefix(trustPinConfig.CA, gun) {
+		trustPinChecker.mode = ca
+		for caGunPrefix, caFilepath := range trustPinConfig.CA {
+			if strings.HasPrefix(gun, caGunPrefix) {
```

Comment:
What happens if there is more than one CA prefix that matches the gun?  Should we pick the most specific one?  Otherwise, since it's looping over a map, it'd be random.

That might be fine, so long as we document what the user can expect the behavior to be.